base/strings/search.jl

```diff
@@ -212,44 +206,52 @@ julia> rsearch("aaabbb","b")
 6:6
 ```
 """
-function rsearch(s::AbstractString, c::Chars, i::Integer)
-    e = endof(s)
-    j = search(RevString(s), c, e-i+1)
-    j == 0 && return 0
-    e-j+1
+function rsearch(s::AbstractString, c::Chars, i::Integer=start(s))
+    @boundscheck checkbounds(s, i)
+    isempty(c) && return i
+    j = Int(i)
+    @inbounds while j > 0
+        s[j] in c && return j
+        j = prevind(s, j)
+    end
+    return 0
 end
 
 function _rsearchindex(s, t, i)
     if isempty(t)
-        return 1 <= i <= nextind(s,endof(s)) ? i :
-               throw(BoundsError(s, i))
+        @boundscheck checkbounds(s, i)
+        return i
     end
-    t = RevString(t)
-    rs = RevString(s)
     l = endof(s)
-    t1, j2 = next(t,start(t))
+    j2 = endof(t)
+    t1 = t[j2]
     while true
-        i = rsearch(s,t1,i)
-        if i == 0 return 0 end
-        c, ii = next(rs,l-i+1)
-        j = j2; k = ii
+        i = rsearch(s, t1, i)
+        i == 0 && return 0
+        c = s[i]
+        ii = prevind(s, i)
+        j, k = j2, ii
         matched = true
-        while !done(t,j)
-            if done(rs,k)
+        while j > 0
+            if k < 1
                 matched = false
                 break
             end
-            c, k = next(rs,k)
-            d, j = next(t,j)
+            # Using `reverseind` with `prevind` mimics `next` but for iteration over
```

Comment:
I still wonder whether it wouldn't be more efficient to work with indices in the original string (using `prevind`), and only call `reverseind` once the match is found. Or maybe I'm missing something?

Reply:
Probably, though this approach didn't introduce any regressions as far as Nanosoldier could tell. My problem with changing it is that it took me quite a while to figure out how to even get to this; translating this algorithm to be in reverse broke my brain a bit. I'd be happy to change it with some guidance though.

Reply:
Yeah maybe just leave a `TODO` so that we can find it when looking for possible performance enhancements.

If you want to try changing it, I'd say you should just have to call `prevind` on both `rs` and `t`, extract the chars and compare them as currently? Actually I don't understand why `reversind` is needed here, nor how it could be correct. In the first iteration, `k` contains the index in `s`, not in `reverse(s)`, so how can that work?

Reply:
I wish I knew.